src/crush/CrushWrapper.cc

```diff
@@ -1817,6 +1821,144 @@ int CrushWrapper::update_device_class(int id,
   return 1;
 }
 
+// Recursively get all only-child shadow parents up to the root
+// based on the given item id.
+// Return -EBUSY if any parent has more than one children.
+// Return -ENOENT if a specific parent can not be found.
+// Return 0 on success.
+int CrushWrapper::get_immediate_only_child_shadow_parents(CephContext *cct,
+                                                          int id,
+                                                          list<int> *parents)
+{
+  assert(parents);
+  parents->clear();
+  while (_search_item_exists(id)) {
```

Comment:
I see.  This goes away if we drop rename, though, right?  That would be my preference.  :)